src/set_theory/zfc/ordinal.lean

```diff
@@ -16,59 +16,127 @@ Further development can be found on the branch `von_neumann_v2`.
+theorem transitive_iff_subset_powerset : x.transitive ↔ x ⊆ powerset x :=
 ⟨λ h y hy, mem_powerset.2 $ h.subset_of_mem hy, λ H y hy z hz, mem_powerset.1 (H hy) hz⟩
 
-alias is_transitive_iff_subset_powerset ↔ is_transitive.subset_powerset _
+alias transitive_iff_subset_powerset ↔ transitive.subset_powerset _
+
+/-! ### Ordinals as sets -/
+
+/-- A set `x` is a von Neumann ordinal when it's a hereditarily transitive set. We prove that this
+further implies that `x` is well-ordered under `∈`. -/
+def is_ordinal (x : Set) : Prop := x.hereditarily transitive
+
+theorem is_ordinal_iff {x : Set} : x.is_ordinal ↔ x.transitive ∧ ∀ y ∈ x, is_ordinal y :=
+hereditarily_iff
+
+alias is_ordinal_iff ↔ is_ordinal.def _
+
+namespace is_ordinal
+
+protected theorem transitive (h : x.is_ordinal) : x.transitive := h.def.1
+
+protected lemma mem (h : x.is_ordinal) {y} (hy : y ∈ x): is_ordinal y := h.def.2 _ hy
+
+theorem subset_of_mem (h : x.is_ordinal) : y ∈ x → y ⊆ x := h.transitive.subset_of_mem
+
+theorem mem_trans (h : z.is_ordinal) : x ∈ y → y ∈ z → x ∈ z := h.transitive.mem_trans
+
+theorem mem_trans' (h : x.is_ordinal) (hw : w ∈ x) : y ∈ z → z ∈ w → y ∈ w := (h.mem hw).mem_trans
+
+protected theorem sUnion' (H : ∀ y ∈ x, is_ordinal y) : (⋃₀ x).is_ordinal :=
+begin
+  refine is_ordinal_iff.mpr ⟨transitive.sUnion' $ λ y hy, (H y hy).transitive, λ y hy, _⟩,
+  obtain ⟨w, hw, hy⟩ := mem_sUnion.1 hy,
+  exact (H _ hw).mem hy,
+end
+
+protected theorem sUnion (H : is_ordinal x) : (⋃₀ x).is_ordinal :=
+is_ordinal.sUnion' (λ y hy, H.mem hy)
+
+protected theorem sUnion_subset (H : is_ordinal x) : ⋃₀ x ⊆ x :=
+λ y hy, let ⟨z, hz, hyz⟩ := mem_sUnion.mp hy in H.mem_trans hyz hz
+
+protected theorem union (hx : x.is_ordinal) (hy : y.is_ordinal) : (x ∪ y).is_ordinal :=
+is_ordinal.sUnion' $ λ z hz, by { rcases mem_pair.1 hz with rfl | rfl; assumption }
+
+protected theorem inter (hx : x.is_ordinal) (hy : y.is_ordinal) : (x ∩ y).is_ordinal :=
+is_ordinal_iff.mpr ⟨hx.transitive.inter hy.transitive, λ z hz, hx.mem (mem_inter.mp hz).1⟩
+
+protected theorem is_trans (h : x.is_ordinal) : is_trans x.to_set (subrel (∈) _) :=
+⟨λ a b c hab hbc, h.mem_trans' c.2 hab hbc⟩
+
+theorem _root_.Set.is_ordinal_iff_mem_transitive (x : Set) :
+  x.is_ordinal ↔ x.transitive ∧ ∀ y ∈ x, transitive y :=
+⟨λ h, ⟨h.transitive, λ y hy, (h.mem hy).transitive⟩, Set.induction_on x (λ x ih ⟨h₁, h₂⟩,
+  is_ordinal_iff.mpr ⟨h₁, λ y hy, ih _ hy ⟨h₂ _ hy, λ z hz, h₂ _ (h₁.mem_trans hz hy)⟩⟩)⟩
+
+theorem _root_.Set.is_ordinal_iff_is_trans (x : Set) :
+  x.is_ordinal ↔ x.transitive ∧ is_trans x.to_set (subrel (∈) _) :=
+⟨λ h, ⟨h.transitive, h.is_trans⟩, λ ⟨h₁, ⟨h₂⟩⟩, x.is_ordinal_iff_mem_transitive.mpr
+  ⟨h₁, λ c hc, transitive_iff_mem_trans.mpr $ λ a b hab hbc, let hb := h₁.mem_trans hbc hc in
+    h₂ ⟨a, h₁.mem_trans hab hb⟩ ⟨b, hb⟩ ⟨c, hc⟩ hab hbc⟩⟩
+
+/-- A relation embedding between a smaller and a larger ordinal. -/
+protected def rel_embedding (hx : x.is_ordinal) (hy : y ∈ x) :
```

Comment:
This feels out of place, since the results that use this aren't in this PR, but I guess it can't hurt.